arcgis-ios-sdk-samples/Layers/Display KML/DisplayKMLViewController.swift

```diff
@@ -0,0 +1,180 @@
+                self?.displayedSource = dataSource
+            }
+        }
+    }
+    
+    //MARK: - KML Loading
+    
+    private func display(kmlLayer: AGSKMLLayer){
+        
+        // Clear the existing layers from the map
+        mapView.map?.operationalLayers.removeAllObjects()
+        // Add the KML layer to the map
+        mapView.map?.operationalLayers.add(kmlLayer)
+        
+        // Show the progress indicator
+        SVProgressHUD.show(withStatus: "Loading KML Layer")
+        
+        // This load call is not required, but it allows for error
+        // feedback and progress indication
+        kmlLayer.load {[weak self] (error) in
+            
+            // Close the progress indicator
+            SVProgressHUD.dismiss()
+
+            if let self = self,
+                let error = error {
+                // Display the error if one occurred
+                let alertController = UIAlertController(title: error.localizedDescription, message: nil, preferredStyle: .alert)
+                alertController.addAction(UIAlertAction(title: "OK", style: .cancel))
+                self.present(alertController, animated: true, completion: nil)
+            }
+        }
+        
+    }
+    
+    private func changeSourceToURL() {
+        /// A URL of a remote KML file
+        let kmlDatasetURL = URL(string: "https://www.wpc.ncep.noaa.gov/kml/noaa_chart/WPC_Day1_SigWx.kml")!
+        let kmlDataset = AGSKMLDataset(url: kmlDatasetURL)
+        /// A KML layer created from a remote KML file
+        let kmlLayer = AGSKMLLayer(kmlDataset: kmlDataset)
+        display(kmlLayer: kmlLayer)
+    }
+    
+    private func changeSourceToLocalFile() {
+        /// A dataset created by referencing the name of a KML file in the app bundle
+        let kmlDataset = AGSKMLDataset(name: "US_State_Capitals")
+        /// A KML layer created from a local KML file
+        let kmlLayer = AGSKMLLayer(kmlDataset: kmlDataset)
+        display(kmlLayer: kmlLayer)
+    }
+    
+    private func changeSourceToPortalItem() {
+        let portal = AGSPortal.arcGISOnline(withLoginRequired: false)
+        /// A remote KML portal item
+        let portalItem = AGSPortalItem(portal: portal, itemID: "9fe0b1bfdcd64c83bd77ea0452c76253")
+        /// A KML layer created from an ArcGIS Online portal item
+        let kmlLayer = AGSKMLLayer(item: portalItem)
+        display(kmlLayer: kmlLayer)
+    }
+    
+}
+
+extension DisplayKMLViewController: UIPopoverPresentationControllerDelegate {
+    
+    func adaptivePresentationStyle(for controller: UIPresentationController) -> UIModalPresentationStyle {
+        // prevent the popover from displaying fullscreen in narrow contexts
+        return .none
+    }
+}
+
+class DisplayKMLDataSourcesTableViewController: UITableViewController {
+    
+    fileprivate var initialSelectedIndex = 0
+    fileprivate var selectionHandler: ((KMLDataSource)->Void)?
```

Comment:
I'd prefer to see this sample follow a delegate design pattern rather than a selection handler.

Reply:
I prefer to use a closure when there's only one callback and it's essentially a completion handler. It's simpler than declaring another type, a function prototype, an implementation, and another object.

Reply:
@qsmmorgan You know where I stand on using delegates. I will say that the more you use them, the easier they become to implement and the more natural they feel.

Reply:
I could say the same about closures haha. If you want me to change it I will.

Reply:
It is certainly easier. I'll suggest the delegate pattern for parent-child view controller relationships is pretty pervasive across cocoa and so it would feel more appropriate to implement rather than a callback. Alternatively, the callback pattern is often used with async completion and this is not that. Also, this places the memory burden on the child view controller, something that isn't needed.